src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -911,4 +925,93 @@ int getBufferedLength() {
             return ibuffer.size();
         }
     }
+
+    /**
+     * Check for overlap. If the src and dst buffers are using shared data and
+     * if dst will overwrite src data before src can be processed.  If so, make
+     * a copy to put the dst data in.
+     */
+    ByteBuffer overlapDetection(ByteBuffer src, ByteBuffer dst) {
+        if (src.isDirect() && dst.isDirect()) {
```

Comment:
originalDst is only set to null at construction time. It should be reset here or in restoreDst(). Otherwise, there may be some strange interaction between various ByteBuffer calls using the same Cipher object. Say, the first call uses an extra copy and set originalDst, then the second call does not, but originalDst is still non-null.

Reply:
Yeah, that makes sense.